presto-cache/src/main/java/com/facebook/presto/cache/AbstractAsyncCacheManager.java

```diff
@@ -0,0 +1,132 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+
+public abstract class AbstractAsyncCacheManager
+        implements CacheManager
+{
+    private static final Logger log = Logger.get(AbstractAsyncCacheManager.class);
+
+    private static final String EXTENSION = ".cache";
+
+    private final ScheduledExecutorService executor = newScheduledThreadPool(5);
+    private final Map<FileCacheKey, Slice> inflight = new ConcurrentHashMap<>();
+    private final Queue<Future<?>> inflightToPersistedJobs = new ConcurrentLinkedQueue<>();
+
+    private final Path directory;
+
+    abstract boolean read(FileCacheKey key, byte[] buffer, int offset);
+
+    abstract boolean write(FileCacheKey key, Slice data, Path path);
+
+    public AbstractAsyncCacheManager(Path directory)
+    {
+        this.directory = requireNonNull(directory, "directory is null");
+
+        File target = new File(directory.toString());
+        if (!target.exists()) {
+            target.mkdir();
+        }
+        else {
+            File[] files = target.listFiles();
+            if (files == null) {
+                return;
+            }
+            Arrays.stream(files).forEach(File::delete);
+        }
+    }
+
+    @Override
+    public boolean get(FileCacheKey key, byte[] buffer, int offset)
+    {
+        drainInflights();
+
+        // There might be a chance the request key is still in memory
+        Slice data = inflight.get(key);
+        if (data != null) {
+            checkState(key.getLength() == data.length(), "cache length mismatch %s vs %s", key.getLength(), data.length());
+            checkState(
+                    offset + key.getLength() <= buffer.length,
+                    "insufficient buffer size %s; requested [%s, %s)",
+                    buffer.length,
+                    offset,
+                    offset + key.getLength());
+
+            data.getBytes(0, buffer, offset, key.getLength());
+            return true;
+        }
+
+        return read(key, buffer, offset);
+    }
+
+    @Override
+    public void put(FileCacheKey key, Slice data)
+    {
+        drainInflights();
+
+        inflight.put(key, data);
+        try {
+            Future<?> future = executor.submit(() -> {
+                Path path = new Path(directory.toString() + "/" + randomUUID() + EXTENSION);
+                if (!write(key, data, path)) {
+                    log.warn("Fail to persist cache %s with length %s ", path, key.getLength());
+                }
+
+                inflight.remove(key);
+            });
+            inflightToPersistedJobs.add(future);
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void drainInflights()
```

Comment:
We probably want to separate scheduled thread to do this instead of draining before every operation, but since it is prototype I guess it's alright.